File: packages/gavio-js/src/interceptors/reliability/fallback.ts
/** fallbackChain (F-REL-02) — route to a secondary provider on failure. */

import type { InterceptorContext } from '../../context.js'
import { ProviderError } from '../../errors.js'
import type { ProviderAdapter } from '../../providers/base.js'
import type { GavioRequest } from '../../request.js'
import type { GavioResponse } from '../../response.js'
import { coerceProvider } from '../../types.js'
import type { Executor, ExecutorPolicy } from '../base.js'

export interface FallbackChainOptions {
  /** Provider adapters to try, in order, when the primary call fails. */
  fallbacks: ProviderAdapter[]
}

/**
 * Try the primary executor; on a provider error, try fallback adapters.
 *
 * Each fallback is a provider adapter (anything with an async `complete`). The
 * request's provider/model are rewritten per fallback so the audit record
 * reflects which provider actually answered.
 */
class FallbackChain implements ExecutorPolicy {
  readonly name = 'fallback'
  readonly isExecutorPolicy = true as const
  readonly dryRunSafe = true

  private readonly fallbacks: ProviderAdapter[]

  constructor(options: FallbackChainOptions) {
    if (!options.fallbacks || options.fallbacks.length === 0) {
      throw new Error('fallbackChain requires at least one fallback adapter')
    }
    this.fallbacks = options.fallbacks
  }

  async around(
    request: GavioRequest,
    ctx: InterceptorContext,
    callNext: Executor,
  ): Promise<GavioResponse> {
    ctx.markFired(this.name)
    try {
      return await callNext(request)
    } catch (primaryError) {
      if (!(primaryError instanceof ProviderError)) throw primaryError
      let lastError: unknown = primaryError
      for (const adapter of this.fallbacks) {
        try {
          const rerouted = request.copyWithMessages(request.messages)
          rerouted.provider = coerceProvider(adapter.providerName)
          return await adapter.complete(rerouted)
        } catch (error) {
          if (!(error instanceof ProviderError)) throw error
          lastError = error
        }
      }
      throw lastError
    }
  }
}

/** Factory: build a fallback policy. */
export function fallbackChain(options: FallbackChainOptions): ExecutorPolicy {
  return new FallbackChain(options)
}

File: packages/gavio-py/gavio/interceptors/reliability/fallback.py
"""FallbackChain (F-REL-02) — route to a secondary provider on failure."""

from __future__ import annotations

import logging

from ...context import InterceptorContext
from ...exceptions import ProviderError
from ...interceptors.chain import Executor
from ...request import GavioRequest
from ...response import GavioResponse
from ...types import Provider
from .policy import ExecutorPolicy

logger = logging.getLogger("gavio.fallback")


class FallbackChain(ExecutorPolicy):
    """Try the primary executor; on a provider error, try fallback adapters.

    Each fallback is a provider adapter (anything with an async ``complete``).
    The request's ``provider``/``model`` are rewritten per fallback so the
    audit record reflects which provider actually answered.
    """

    def __init__(self, fallbacks: list) -> None:
        if not fallbacks:
            raise ValueError("FallbackChain requires at least one fallback adapter")
        self.fallbacks = fallbacks

    @property
    def name(self) -> str:
        return "fallback"

    async def around(
        self,
        request: GavioRequest,
        ctx: InterceptorContext,
        call_next: Executor,
    ) -> GavioResponse:
        ctx.mark_fired(self.name)
        try:
            return await call_next(request)
        except ProviderError as primary_error:
            logger.warning(
                "fallback: primary failed (%s); trying %d fallback(s)",
                type(primary_error).__name__,
                len(self.fallbacks),
            )
            last_error: Exception = primary_error
            for adapter in self.fallbacks:
                try:
                    rerouted = request.copy_with_messages(request.messages)
                    rerouted.provider = Provider.coerce(adapter.provider_name)
                    return await adapter.complete(rerouted)
                except ProviderError as error:
                    last_error = error
                    logger.warning(
                        "fallback: %s also failed (%s)",
                        adapter.provider_name,
                        type(error).__name__,
                    )
            raise last_error from primary_error
